cabal-install/src/Distribution/Client/FetchUtils.hs

```diff
@@ -118,6 +122,34 @@ checkRepoTarballFetched repo pkgid = do
       then return (Just file)
       else return Nothing
 
+verifyFetchedTarball :: Verbosity -> RepoContext -> Repo -> PackageId -> IO Bool
+verifyFetchedTarball verbosity repoCtxt repo pkgid =
+   let file = packageFile repo pkgid
+       handleError :: IO Bool -> IO Bool
+       handleError act = do
+         res <- Safe.try act
+         case res of
+           Left e -> warn verbosity ("Error verifying fetched tarball " ++ file ++ ", will redownload: " ++ show (e :: SomeException)) >> pure False
```

Comment:
Is this error message triggered if `file` simply does not exist?

Reply:
Yes, this function `verifyFetchedTarball` is called only in one place in the code, for all elements of `repoTarballPkgsWithMetadataUnvalidated`, which seems to contain ids of any packages that come from a secure repo, regardless of whether they were ever downloaded:

https://github.com/haskell/cabal/pull/8500/files#diff-2adae381a2eb2d8c804d5396694a431e1e3babd816a5d2408a62213cf774a893R929-R935

Reply:
A solution would be to move the partitioning wrt if a package is already downloaded (the block containing `checkRepoTarballFetched`) earlier and apply it to `allPkgLocations` instead of `repoTarballPkgsWithoutMetadata`. Only packages that are already downloaded would then be checked for `RepoSecure` and then verified. Does that make sense?

Reply:
Or, better, avoiding some filesystem IO, partition `repoTarballPkgsWithMetadataUnvalidated` wrt if a package is already downloaded.